### DISCUSSION TITLE
Dynamic/adaptable contours in matlab

### DISCUSSION THREAD

**Discussion by AlexPuckett** [community/community#142630] [Programming Help]
### Body

Hello everyone,
So i have been trying to create a function in matlab that will plot contours on a floor plan with detected walls and adapt along them. i.e. decreasing exponentially when intersecting with a wall, but i can not find a way to resolve this. I would appreciate your help.
This is the code for the function
and an image of what i am currently getting.
You can find the entire code at my repository Thesiscopy3.
![untitled](https://github.com/user-attachments/assets/4a497272-49c8-4138-bd66-4ee4777e7a4f)
`% Function to detect walls in the image and plot them in 2D
function visualizeWalls2D(lengthField, widthField, distanceEditField, sourceDropdown, sourceData, doseEditField, rateEditField, activityEditField, density, tableData)
    % Retrieve source data
    selectedSource = sourceData.(sourceDropdown.Value);
    materials = fieldnames(selectedSource.TVLe);

    % Retrieve the image from the application data
    img = getappdata(0, 'FloorPlanImage');
    if isempty(img)
        return;
    end

    % Get the user-defined Length, Width, and Thickness
    floorLength = lengthField.Value;
    floorWidth = widthField.Value;

    if isempty(floorLength) || isempty(floorWidth)
        uialert(gcf, 'Please enter valid dimensions!', 'Invalid Input');
        return;
    end

    % Convert image to grayscale and adjust contrast
    gs = im2gray(img);
    gs = imadjust(gs);

    % Detect edges using edge detection (e.g., Canny method)
    edges = edge(gs, 'Canny', [0.1 0.3]);

    % Find the boundaries of the objects in the binary image
    [B, ~] = bwboundaries(edges, 'noholes');

    for i = 1:length(materials)
        material = materials{i};
        % Create a new figure for the 2D visualization
        fig = figure("Name", "2D " + material + " Wall Visualization");
        ax = axes('Parent', fig); % Create axes in the new figure
        view(ax, 2); % Set view to 2D
        xlabel(ax, 'X');
        ylabel(ax, 'Y');

        hold(ax, 'on'); % Hold the plot for multiple walls
        for k = 1:length(B)
            boundary = B{k};

            % Normalize the boundary coordinates to be between 0 and 1
            normalized_x = boundary(:,2) / size(edges, 2); % Normalize by image width
            normalized_y = 1 - (boundary(:,1) / size(edges, 1)); % Normalize by image height

            % Scale normalized coordinates based on the user input
            scaled_x = normalized_x * floorLength;
            scaled_y = normalized_y * floorWidth;

            % Plot the wall with user-defined thickness
            plot(ax, scaled_x, scaled_y, 'k', 'LineWidth', 2); % boundaries
            hold on
        end
        
        distance = zeros(1,4);
        sourceX = distanceEditField{1}.Value;
        sourceY = distanceEditField{2}.Value;
        plot(ax, sourceX, sourceY, 'r*', 'MarkerSize', 8);
        hold on
        for j = 1:4
            distance(j) = distanceEditField{j}.Value;
        end
        
        [maxdistance, idx] = max(distance);
        [r, th] = meshgrid(linspace(0.25, maxdistance, 200), linspace(0, 2*pi, 200));
        rx = r.*cos(th); % Distance r of dose point from axis origin
        ry = r.*sin(th);
        dw_s = sqrt((scaled_x - sourceX).^2 + (scaled_y - sourceY).^2);
        
        dosedistribution = zeros(size(r));
        for l = 1:numel(r)
            if any(r(l) >= dw_s)
                % Apply attenuation if within wall boundaries
                dosedistribution(l) = (activityEditField.Value * selectedSource.RAKR * doseEditField.Value) ./ (r(l).^2 * rateEditField.Value * 60) * exp(-selectedSource.(material) * density.(material) * tableData{i, 2*idx});
            else
                % Normal dose calculation without attenuation
                dosedistribution(l) = (activityEditField.Value * selectedSource.RAKR * doseEditField.Value) ./ (r(l).^2 * rateEditField.Value * 60);
            end
        end
        
        maxDose = max(dosedistribution(:), [], 'omitnan');
        contourLevels = logspace(log10(maxDose/100), log10(maxDose), 20); % 20 contour levels with closer spacing near lower doses
        
        % Plot the dose distribution contours
        contour(ax, sourceX + rx, sourceY + ry, dosedistribution, contourLevels, 'ShowText', 'on');
        colorbar;
        hold(ax, 'off');
    end
end`

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by VishalKotekar** [comment]
Here’s a more humanized response for the discussion:

Hey Alex,

That’s a really interesting project you’re working on! For making those contours adapt dynamically to the walls, you might try using the inpolygon function in MATLAB. It can help you check if your contours intersect with wall boundaries, allowing you to apply a decay effect when they do.

If you need help tweaking the code or further ideas, feel free to share what you have—I’d love to assist!

Looking forward to seeing what you come up with!

**Reply by AlexPuckett** [reply]
Hello VishalKotekar,

Thank you for your response I will give it a try. I also thought of solving it using some region growing algorithm, but I think your solution might be simpler. What are your thoughts on the use of a region growing algorithm?
